solutions/cses-1753.mdx

```diff
@@ -0,0 +1,161 @@
+---
+id: cses-1753
+source: CSES
+title: String Matching
+author: Dustin Miao
+---
+
+## Solution - Knuth-Morris-Prat Algorithm
+
+Define an array $\pi_S$ on some string $S$ such that $\pi_S[i]$ stores the length of the longest nontrivial prefix of the entire string that is equivalent to a suffix ending at position $i$. Formally, 
+
+$$
+\pi_S[i] = \max\{k \: | \: 1 \leq k < i \text{ and } S[0:k - 1] \equiv S[i - (k - 1): i]\}
+$$
+
+If we are searching for string $P$ inside of string $T$, create a new string $S = P + \# + T$, where $\#$ is any arbitrary character that does not occur in either string. Then, create an array $\pi_S$ using the [KMP](https://cp-algorithms.com/string/prefix-function.html) algorithm. The answer is simply the number of indicies inside of $\pi_S$ that is equal to $|P|$ (the length of $P$). 
+
+<LanguageSection>
+
+<CPPSection>
+
+**Time Complexity:** $\mathcal{O}(n + m)$
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+namespace str {
```

Comment:
namespace needed?

Reply:
yes, imo it makes it more clear
~~also i just black box all the string algos~~

Reply:
what? how does it make it more clear?

Reply:
because now you know its a string algorithm
also i use `z` as a variable name in the main function

Reply:
...
dustin, it's _obviously_ a string algorithm i don't know why that needs to be made clear

Reply:
there's no harm in doing so

Reply:
![image](https://user-images.githubusercontent.com/55369003/142446110-19f765ee-bb65-48e7-9700-274fa403d474.png)
https://stackoverflow.com/questions/603378/c-namespace-usage-and-naming-rules

Reply:
its literally for naming conflicts

Reply:
oh wait
i'm bad
sorry
yeah, carry on